### PR TITLE
feat: improve blockchain api

### PR DESCRIPTION
**Pull request by sokel**
- Do  Events API more flexible

- Add methods to pack and unpack transaction data for MultiSig usage

- Add MultisigAPI

### CODE REVIEW

File: blockchain/api.go [hunk 1]
```diff
@@ -137,11 +143,33 @@ type SimpleGatekeeperAPI interface {
...
 
+type MultiSigAPI interface {
+	AddOwner(ctx context.Context, key *ecdsa.PrivateKey, owner common.Address) error
+	RemoveOwner(ctx context.Context, key *ecdsa.PrivateKey, owner common.Address) error
+	ReplaceOwner(ctx context.Context, key *ecdsa.PrivateKey, owner common.Address, newOwner common.Address) error
+	ChangeRequirement(ctx context.Context, key *ecdsa.PrivateKey, required *big.Int) error
+	SubmitTransaction(ctx context.Context, key *ecdsa.PrivateKey, destination common.Address, value *big.Int, data []byte) error
+	ConfirmTransaction(ctx context.Context, key *ecdsa.PrivateKey, transactionId *big.Int) error
+	RevokeConfirmation(ctx context.Context, key *ecdsa.PrivateKey, transactionId *big.Int) error
+	ExecuteTransaction(ctx context.Context, key *ecdsa.PrivateKey, transactionId *big.Int) error
```
Comment (sshaman1101) [inline blockchain/api.go]: Id -> ID

File: blockchain/api.go [hunk 2]
```diff
@@ -49,7 +51,7 @@ type ProfileRegistryAPI interface {
 }
 
 type EventsAPI interface {
-	GetEvents(ctx context.Context, fromBlockInitial *big.Int) (chan *Event, error)
+	GetEvents(ctx context.Context, filter *EventFilter, fromBlockInitial *big.Int) (chan *Event, error)
```
Comment (sshaman1101) [inline blockchain/api.go]: maybe also move fromBlock into filters

File: blockchain/api.go [hunk 3]
```diff
@@ -1322,6 +1446,31 @@ func (api *OracleUSDAPI) SetCurrentPrice(ctx context.Context, key *ecdsa.Private
...
+
+	if len(data) < 36 {
+		return nil, fmt.Errorf("data is malformed")
+	}
+	// Cut method specify byte, stands only parameters data
+	// Method ids are saves at the first 4 bytes of the hash of the
+	// methods string signature. (signature = baz(uint32,string32))
+	params := data[4:]
+
+	price := common.HexToAddress(common.Bytes2Hex(params[:32])).Big()
```
Comment (sshaman1101) [inline blockchain/api.go]: why not using just `data[4:32]`?

File: blockchain/api.go [hunk 4]
```diff
@@ -1369,7 +1518,195 @@ func (api *BasicSimpleGatekeeper) Payout(ctx context.Context, key *ecdsa.Private
...
+
+	return gateABI.Pack("Payout", to, value, txNumber)
+}
+
+func (api *BasicSimpleGatekeeper) UnpackPayoutTransactionData(data []byte) (*SimpleGateTransaction, error) {
+	if len(data) == 0 {
+		return nil, fmt.Errorf("data is empty")
+	}
+
+	if len(data) < 100 {
```
Comment (sshaman1101) [inline blockchain/api.go]: You're requiring len=100 but below using only 96 bytes. WTF?

File: blockchain/api.go [hunk 5]
```diff
@@ -1369,7 +1518,195 @@ func (api *BasicSimpleGatekeeper) Payout(ctx context.Context, key *ecdsa.Private
...
+func (api *BasicMultiSigAPI) GetTransactionIds(ctx context.Context, from *big.Int, to *big.Int, pending bool, executed bool) ([]*big.Int, error) {
+	return api.contract.GetTransactionIds(getCallOptions(ctx), from, to, pending, executed)
+}
+
+func (api *BasicMultiSigAPI) GetTransaction(ctx context.Context, transactionId *big.Int) (*MultiSigTransactionData, error) {
+	tx, err := api.contract.Transactions(getCallOptions(ctx), transactionId)
+	if err != nil {
+		return nil, err
+	}
+	return &MultiSigTransactionData{
```
Comment (sshaman1101) [inline blockchain/api.go]: that the reason for introducing new struct and not using `types.Transaction`?

File: blockchain/api.go [hunk 6]
```diff
@@ -126,6 +128,10 @@ type OracleAPI interface {
 	SetCurrentPrice(ctx context.Context, key *ecdsa.PrivateKey, price *big.Int) (*types.Transaction, error)
 	// GetCurrentPrice returns current price relation between some currency and SONM token
 	GetCurrentPrice(ctx context.Context) (*big.Int, error)
+	// PackSetCurrentPriceTransactionData pack `SetCurrentPrice` method call
+	PackSetCurrentPriceTransactionData(price *big.Int) ([]byte, error)
+	// UnpackSetCurrentPriceTransactionData unpack `SetCurrentPrice` method call
+	UnpackSetCurrentPriceTransactionData(data []byte) (*big.Int, error)
```
Comment (sshaman1101) [inline blockchain/api.go]: Seems like this method is related to data-representation. Should it exists as part of the interface, or may be better to leave it as not exportable part of implementation?

File: blockchain/api.go (hunk 5 above)
Comment (sokel) [reply blockchain/api.go]: its not just blockchain tx

File: blockchain/api.go (hunk 4 above)
Comment (sokel) [reply blockchain/api.go]: 4 byte for method definition
3x32 byte for parameters

File: blockchain/api.go (hunk 3 above)
Comment (sokel) [reply blockchain/api.go]: its just for separate method and parameters.
more visual effect, for developers. =)

File: blockchain/api.go (hunk 2 above)
Comment (sokel) [reply blockchain/api.go]: good idea for next fix.

File: blockchain/api.go [hunk 7]
```diff
@@ -137,11 +143,33 @@ type SimpleGatekeeperAPI interface {
 	// Payout release payout transaction from mirrored chain.
 	// Accessible only by owner.
 	Payout(ctx context.Context, key *ecdsa.PrivateKey, to common.Address, value *big.Int, txNumber *big.Int) (*types.Transaction, error)
+	// PackPayoutTransactionData pack `Payout` method call
+	PackPayoutTransactionData(to common.Address, value *big.Int, txNumber *big.Int) ([]byte, error)
```
[the remark this answers is not in the document]
Comment (sokel) [reply blockchain/api.go]: you right, thx

File: blockchain/api.go (hunk 4 above)
Comment (sshaman1101) [reply blockchain/api.go]: oh, you're cutting the header, got it.

File: blockchain/api.go (hunk 5 above)
Comment (sshaman1101) [reply blockchain/api.go]: but it looks like just a subset of fields from `types.Transaction`, isn't it?

File: blockchain/api.go (hunk 2 above)
Comment (sshaman1101) [reply blockchain/api.go]: next fix, seriously? You're just peretrakhal all of the filters, why not just peretrakhat' this one?

Comment (sokel) [reply blockchain/api.go]: you inspired me to do this =)

File: blockchain/api.go (hunk 6 above)
Comment (sokel) [reply blockchain/api.go]: i have are plan.

Comment (sshaman1101) [reply blockchain/api.go]: Sooooo, can you dedicate us into your plan?

**Pull request closed without merging**